<br>

# Research: Assessing the reliability of borrowers of bank


***

**Customer** - credit department of the bank. You need to find out whether the marital status and the number of children of the client affects the fact of repayment of the loan on time. Input data from the bank - statistics on the solvency of customers.

The results of the study will be taken into account when building the model **of credit scoring** - a special system that evaluates the ability of a potential borrower to repay a loan to a bank.

**Data Description:**
- children - the number of children in the family
- days_employed - total work experience in days
- dob_years - customer age in years
- education - customer education level
- education_id - educational level identifier
- family_status - marital status
- family_status_id - identifier of marital status
- gender - customer gender
- income_type - type of employment
- debt - whether there was a loan repayment debt
- total_income - monthly income
- purpose - the purpose of obtaining a loan.
<br>
<br>

***

## Step 1. Data Requirement Gathering

### Data Learning Sequence:
1. Connection of the pandas library
2. Read the data from the data.csv file
3. View the size of the data table
4. View general data information in data
5. Viewing the first 10 rows of the data table
6. View the last 10 rows of the data table
7. View a list of column names
8. Search and view the number of duplicates in the data dataset
9. Search and view the number of empty values in the data set <br>
**We study the data for each of the columns:**
10. Data column children. Let's calculate the number of children in the families of borrowers of the bank
11. Data column days_employed. We will calculate the total number of days of experience of borrowers
12. Data column days_employed. Let's look at the data display in the column
13. The data in the dob_years column. We will calculate the total number of days of borrower experience in years
14. Data column education. Let's count the number of borrowers with different levels of education
15. Data column family_status. Let's count the number of borrowers with different marital status
16. Data column gender. Let's count the number of borrowers of different sexes
17. Data column income_type. Let's count the number of borrowers of different types of employment
18. Data column total_income. Let's see if there are borrowers without income
19. Data column total_income. We will calculate the monthly income of all borrowers
20. Data column total_income. Let's see the missing values
21. Data column purpose. Let's see the number of different goals of borrowers <br>
**Conclusion on the studied data**

In [2]:
import pandas as pd

In [5]:
data = pd.read_csv('data.csv')

<span style="color:blue">View data table size.</span>

In [6]:
data.shape

(21525, 13)

<span style="color:blue">View general data information in data</span>

In [7]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
Unnamed: 0          21525 non-null int64
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB
None


In [8]:
data.head(10)

Unnamed: 0  children  days_employed  dob_years education  education_id  \
0           0         1   -8437.673028         42    высшее             0   
1           1         1   -4024.803754         36   среднее             1   
2           2         0   -5623.422610         33   Среднее             1   
3           3         3   -4124.747207         32   среднее             1   
4           4         0  340266.072047         53   среднее             1   
5           5         0    -926.185831         27    высшее             0   
6           6         0   -2879.202052         43    высшее             0   
7           7         0    -152.779569         50   СРЕДНЕЕ             1   
8           8         2   -6929.865299         35    ВЫСШЕЕ             0   
9           9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [9]:
data.tail(10)

Unnamed: 0  children  days_employed  dob_years       education  \
21515       21515         1    -467.685130         28         среднее   
21516       21516         0    -914.391429         42          высшее   
21517       21517         0    -404.679034         42          высшее   
21518       21518         0  373995.710838         59         СРЕДНЕЕ   
21519       21519         1   -2351.431934         37  ученая степень   
21520       21520         1   -4529.316663         43         среднее   
21521       21521         0  343937.404131         67         среднее   
21522       21522         1   -2113.346888         38         среднее   
21523       21523         3   -3112.481705         38         среднее   
21524       21524         2   -1984.507589         40         среднее   

       education_id     family_status  family_status_id gender income_type  \
21515             1   женат / замужем                 0      F   сотрудник   
21516             0   женат / замужем                 0      F   компаньон   
21517             0  гражданский брак                 1      F   компаньон   
21518             1   женат / замужем                 0      F   пенсионер   
21519             4         в разводе                 3      M   сотрудник   
21520             1  гражданский брак                 1      F   компаньон   
21521             1   женат / замужем                 0      F   пенсионер   
21522             1  гражданский брак                 1      M   сотрудник   
21523             1   женат / замужем                 0      M   сотрудник   
21524             1   женат / замужем                 0      F   сотрудник   

       debt   total_income                            purpose  
21515     1  109486.327999              заняться образованием  
21516     0  322807.776603               покупка своего жилья  
21517     0  178059.553491       на покупку своего автомобиля  
21518     0  153864.650328               сделка с автомобилем  
21519     0  115949.039788  покупка коммерческой недвижимости  
21520     0  224791.862382                  операции с жильем  
21521     0  155999.806512               сделка с автомобилем  
21522     1   89672.561153                       недвижимость  
21523     1  244093.050500       на покупку своего автомобиля  
21524     0   82047.418899              на покупку автомобиля

<span style="color:blue">View at the list of column names, if there are no spaces in the column names, then I won’t rename the column names.</span>

In [10]:
data.columns

Index(['Unnamed: 0', 'children', 'days_employed', 'dob_years', 'education',
       'education_id', 'family_status', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose'],
      dtype='object')

<span style="color:blue">Let's see if there are duplicates in the table and print their number.</span>

In [11]:
data.duplicated().sum()

0

<span style="color:blue">Let's see if there are gaps in the table and calculate their number.</span>

In [40]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<span style="color:blue">Let's see the data in the children column. We calculate the number of children in the families of the borrowers of the bank.</span>

In [41]:
data.sort_values('children')['children'].value_counts().sort_index()

-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

<span style="color:blue">Let's look at the data of the days_employed column. We will calculate the total number of days of experience of borrowers.</span>

In [42]:
data[data['days_employed'] < 0]['days_employed'].count()

15906

<span style="color:blue">Let's look at the data of the days_employed column. Let's look at the data display in the column.</span>

In [43]:
data[data['days_employed'] > 0].head()

children  days_employed  dob_years education  education_id  \
4          0  340266.072047         53   среднее             1   
18         0  400281.136913         53   среднее             1   
24         1  338551.952911         57   среднее             1   
25         0  363548.489348         67   среднее             1   
30         1  335581.668515         62   среднее             1   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   
30        женат / замужем                 0      F   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости  
30  171456.067993  операции с коммерческой недвижимостью

<span style="color:blue">Let's see the data in the dob_years column. We will calculate the total number of days of borrower experience in years.</span>

In [44]:
data.sort_values('dob_years')['dob_years'].value_counts().sort_index()

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

<span style="color:blue">Let's look at the data in the education column. Let's count the number of borrowers with different levels of education.</span>

In [45]:
data.sort_values('education')['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

<span style="color:blue">Let's look at the data in the family_status column. Let's count the number of borrowers with different marital status.</span>

In [46]:
data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

<span style="color:blue">Let's look at the data in the gender column. Let's count the number of borrowers of different sexes.</span>

In [47]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

<span style="color:blue">Let's look at the data in the income_type column. Let's count the number of borrowers of different types of employment.</span>

In [48]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

<span style="color:blue">Let's look at the data of the column total_income. Let's see if there are borrowers without income.</span>

In [49]:
data[data['total_income'] == 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

<span style="color:blue">Let's look at the data of the column total_income. We will calculate the monthly income of all borrowers.</span>

In [50]:
data['total_income'].count()

19351

<span style="color:blue">Let's look at the data of the column total_income. Let's see the missing values.</span>

In [51]:
data[data['total_income'].isnull()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

<span style="color:blue">Let's see the data in the purpose column. Let's see the number of different loan objectives.</span>

In [52]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

### Сonclusion: Step 1

* In the data set data 21 525 records, 12 columns;
* Duplicates in the data set - 54;
* Children column:
    * 47 records (information on the number of children of borrowers) in the column children = -1.
    * These data must be deleted, as they will not give us the correct results;
    * 76 entries (information on the number of children of borrowers) in the column children = 20.
    * This is most likely a mistake, this data will also be deleted from the analysis;
* Days_employed column:
    * 2 174 passes in the days_employed and total_income columns;
    * In the days_employed column, the data type is float64, the number of negative values ​​in the column = 15906.
    * It is necessary to change the data type in the column to int64 and make absolute values;
* Dob_years column:
    * 101 records (borrower age) in the dob_years = 0 column.
* Education column:
    * In the column education there are categories written in a different register, it is necessary to bring to a single look.
* Column family_status:
    * In the column family_status one category it is necessary to bring the writing of all categories to a single form.
* Purpose column:
    * In the purpose column, there are various formulations of the same goals, it is necessary to lead to a uniform view by applying lemmatization.

Assumption of gaps and negative data values:
* There are 76 entries in the children column (information on the number of children of borrowers) in the children = 20. column. This is extremely suspicious, I would like to clarify with the person who provided me the data, which means the value 20. My guess about this value was that the children 2 and 0 it was added by mistake, but again I can’t clarify this - therefore deleted 76 records;
* In the days_employed and total_income columns, the same amount of missing data = 2 174. In my opinion, these columns are interconnected, because on the basis of days_employed we calculate total_income. The reasons for missing data in these columns may depend on the employer who provides data on the number of days worked. And here the employer may not be honest and enter one in the work book and in the statements, and tell the employee another.

<br>
<br>
<br>

***

## Step 2. Data Processing

### Processing passes

In [53]:
import numpy as np

<span style="color:blue">We display the data in the pivot table for 4 columns.</span>

In [55]:
median_values = pd.pivot_table(data,
    index = ['education', 'income_type'],
    values = ['days_employed', 'total_income'],
    aggfunc = np.median
)

median_values

days_employed   total_income
education           income_type                                  
ВЫСШЕЕ              госслужащий       -1424.562274  174909.179988
                    компаньон         -1423.978136  200701.997023
                    пенсионер        367532.282115  165563.164062
                    сотрудник         -1602.753306  144683.271465
Высшее              безработный      395302.838654  202722.511368
                    госслужащий       -3750.610464  161671.347625
                    компаньон         -1453.451901  192451.667511
                    пенсионер        368379.850772  154962.528710
                    сотрудник         -1534.735406  178238.955336
НАЧАЛЬНОЕ           компаньон         -1265.943306  190952.908390
                    пенсионер        360264.985350   90479.336389
                    сотрудник         -1012.313022  212545.366151
НЕОКОНЧЕННОЕ ВЫСШЕЕ компаньон          -775.760267  145528.679914
                    пенсионер        358732.475441  412386.347106
                    сотрудник         -1546.148345  166015.790504
Начальное           госслужащий        -301.333707  190912.178349
                    компаньон         -1030.219648   88227.830722
                    пенсионер        364632.447491   98853.333449
                    сотрудник          -852.697747  129895.001992
Неоконченное высшее госслужащий       -1178.018163  107224.860090
                    компаньон          -711.837088  183379.320368
                    пенсионер        396568.887159   83625.583859
                    сотрудник         -1653.837028  159115.726175
СРЕДНЕЕ             в декрете         -3296.759962   53829.130729
                    госслужащий       -3356.097160  132756.053048
                    компаньон         -1945.914306  157308.365600
                    пенсионер        363541.465259  113656.027279
                    сотрудник         -1567.112656  135823.934197
Среднее             госслужащий       -2291.747052  148529.629601
                    компаньон         -1863.730566  149190.998579
                    пенсионер        363995.316389  128155.229215
                    сотрудник         -1730.667138  133399.107243
УЧЕНАЯ СТЕПЕНЬ      сотрудник          -409.200149  198570.757322
Ученая степень      пенсионер        337584.815560   98752.495442
высшее              госслужащий       -2577.299850  172511.107016
                    компаньон         -1461.167287  201911.716329
                    пенсионер        366157.236636  142825.046222
                    предприниматель    -520.848083  499163.144947
                    сотрудник         -1548.646517  166164.078024
                    студент            -578.751554   98201.625314
начальное           госслужащий       -3043.933615  105766.403301
                    компаньон         -1166.810199  170077.891115
                    пенсионер        358777.901745  104505.757848
                    сотрудник         -1261.719742  124298.759469
неоконченное высшее госслужащий       -2038.132156  165106.669771
                    компаньон         -1032.968329  179867.152890
                    пенсионер        371193.915404  122706.716059
                    сотрудник         -1158.032596  149551.191876
среднее             безработный      337524.466835   59956.991984
                    госслужащий       -2856.781570  136652.970357
                    компаньон         -1645.331963  159652.900458
                    пенсионер        365232.442222  114483.373934
                    сотрудник         -1607.383724  136818.115423
ученая степень      госслужащий       -5968.075884  111392.231107
                    пенсионер        376276.219531  255425.196556
                    сотрудник         -3851.735057  192180.127162

<span style="color:blue">Need a function to fill in the missing values of the days of the experience of borrowers<br>
In the gaps, we need to substitute the median value that we found using median_values.</span>

In [57]:
def fill_days_employed(row):
    return median_values.loc[row['education']].loc[row['income_type']]['days_employed']

<span style="color:blue">In the values of the days_employed column, where there are gaps, we substitute the calculated values.</span>

In [59]:
data.loc[
    data['days_employed'].isnull(), 'days_employed'
] = data.apply(fill_days_employed, axis=1)

<span style="color:blue">Need a function to fill in the missing values of the income level of borrowers.<br>
In the gaps, we need to substitute the median value that we found using median_values.</span>

In [61]:
def fill_total_income(row):
    return median_values.loc[row['education']].loc[row['income_type']]['total_income']

<span style="color:blue">In the values of the column total_income, where there are gaps, we substitute the calculated values.</span>

In [63]:
data.loc[
    data['total_income'].isnull(), 'total_income'
] = data.apply(fill_total_income, axis=1)

<span style="color:blue">It is necessary to get rid of records that have invalid data.</span>

In [65]:
data = data[(data['children'] >= 0) & (data['children'] <= 5)]
data.sort_values('children')['children'].value_counts().sort_index()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

### Сonclusion: Processing passes

<span style="color:blue">Let's see the availability of passes.</span>

In [67]:
data.shape[0]

21402

* No data gaps
* Found the median value and substituted it into the gaps
* Now there are no gaps in the data, we can further analyze them
<br>
<br>

***

### Replacing data types

<span style="color:blue">To translate the values of the days_employed column from float64 to int, use the method.<br>
Converted the number of days of service into an integer type using the astype method.</span>

In [69]:
data['days_employed'] = data['days_employed'].astype('int').apply(abs)

<span style="color:blue">To translate the values of the total_income column from type float64 to int, we use the astype method.<br>
In my opinion, the astype method will be better, since we will not have a floating point after the value.<br>
Nevertheless, I will write down another way how we could replace the data type.<br>
data ['total_income'] = pd.to_numeric (data ['total_income'])</span>

In [71]:
data['total_income'] = data['total_income'].astype('int').apply(abs)

In [72]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   СРЕДНЕЕ             1   
8         2           6929         35    ВЫСШЕЕ             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Сonclusion: Replacing data types

* Converted the values of the days_employed column from float64 to int using the astype method
* We converted the values of the total_income column from type float64 to int using the astype method, did not use the to_numeric () method, since as a result we would have a floating-point number as in the original data.
* Other data types would not suit us in this case, for the reason that the days_employed column had negative values, and now the int data type corrected them
<br>
<br>

***

### Duplicate Processing

<span style="color:blue">You must cast the values in the education column to the same register type.<br>
We use the str.lower () method to translate all the data into string values.</span>

In [73]:
data['education'] = data['education'].str.lower()

<span style="color:blue">Let's see the total number of borrowers with different levels of education.</span>

In [74]:
data.sort_values('education')['education'].value_counts()

среднее                15136
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

<span style="color:blue">Delete the same rows in the data array.</span>

In [134]:
data = data.drop_duplicates().reset_index(drop = True)

<span style="color:blue">Let's see the total number of duplicates.</span>

In [136]:
data.duplicated(keep=False).reset_index(drop=True).sum()

0

### Сonclusion: Duplicate Processing

* Converted values in the education column to one register
* Using the duplicated () method, we calculated the number of duplicates, rows with duplicate values in indexes 106
* Removed all the same lines
<br>
<br>

***

### Lemmatization

<span style="color:blue">Counting the various values in a list.<br>
We use the Counter collection, which implements a dictionary for counting the number of immutable objects.</span>

In [77]:
from collections import Counter

<span style="color:blue"># Import the Mystem library.<br>
We get stemmers / lemmatizers for words in Russian.</span>

In [78]:
from pymystem3 import Mystem
m = Mystem()

<span style="color:blue">It is necessary to find the most frequently occurring words in lemmas.<br>
Based on the search results for frequently occurring words, create loan goal categories.</span>

In [79]:
purposes = []

for p in data['purpose']:
    lemma = m.lemmatize(p)
    purposes += filter(lambda l: l not in [' ', '\n', 'с', 'со', 'на'], lemma)

print(Counter(purposes))

Counter({'недвижимость': 6330, 'покупка': 5880, 'жилье': 4450, 'автомобиль': 4288, 'образование': 3997, 'операция': 2593, 'свадьба': 2337, 'свой': 2224, 'строительство': 1870, 'высокий': 1368, 'получение': 1311, 'коммерческий': 1307, 'для': 1291, 'жилой': 1225, 'сделка': 939, 'заниматься': 908, 'дополнительный': 902, 'проведение': 772, 'сыграть': 769, 'сдача': 651, 'семья': 640, 'собственный': 628, 'ремонт': 609, 'подержанный': 484, 'подержать': 472, 'приобретение': 460, 'профильный': 432})


<span style="color:blue">We saw the words most often found.<br>
Create categories based on common words.</span>

In [80]:
purpose_category = [
    'свадьба',
    'недвижимость',
    'жилье',
    'автомобиль',
    'образование'
]

<span style="color:blue">We will write a function that will lemmatize strings with Russian text and determine a category based on the accounting of a word inside a phrase.</span>

In [81]:
def categorize_purpose(text):
    
    lemmas = m.lemmatize(text)
    for lemma in lemmas:
        for category in purpose_category:
            if category in lemma:
                return category
        

categorize_purpose('На приобретение жилья')

'жилье'

<span style="color:blue">Add a new purpose_category column that defines the categories of goals.</span>

In [82]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [83]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category  
0               покупка жилья            жилье  
1     приобретение автомобиля       автомобиль  
2               покупка жилья            жилье  
3  дополнительное образование      образование  
4             сыграть свадьбу          свадьба  
5               покупка жилья            жилье  
6           операции с жильем            жилье  
7                 образование      образование  
8       на проведение свадьбы          свадьба  
9     покупка жилья для семьи            жилье

<span style="color:blue">We calculate the number of borrowers with specific goals for obtaining a loan.</span>

In [84]:
data.sort_values('purpose_category')['purpose_category'].value_counts()

недвижимость    6330
жилье           4450
автомобиль      4288
образование     3997
свадьба         2337
Name: purpose_category, dtype: int64

### Сonclusion: Lemmatization

* Using lemmatization, we received 5 categories of goals of borrowers to obtain a loan
* Added a column with the loan purpose category, now you can more clearly see the loan objectives of borrowers
* Calculated the number of borrowers with specific goals for obtaining a loan
* To search for values and categorize them, I used the lemmatize () method in which I wrote a function that lemmatized strings with Russian text and determined a category based on the inclusion of a word inside a phrase. Then returns the title categories.
<br>
<br>
<br>

***

## Step 3. Data Questions

### Question 1
- **Is there a relationship between having children and repaying a loan on time?**

In order to find the relationship between the presence of children from borrowers and repayment of the loan on time, it is necessary to segment the borrowers by the presence and number of children.

That is, we need to categorize the data on the column children and debt. <br>
So we categorize the data on the presence of children from borrowers and the presence of debt for repaying a loan.

<span style="color:blue">Distribute borrowers by the presence of children.</span>

In [85]:
def categorize_children(count):
    categories = {
        0: 'бездетная',
        1: 'однодетная',
        2: 'малодетная'
    }

    return categories.get(count, 'многодетная')

<span style="color:blue">Applying the apply () data categorization method to values in the children column.</span>

In [86]:
data['children_category'] = data['children'].apply(categorize_children)

In [87]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose purpose_category children_category  
0               покупка жилья            жилье        однодетная  
1     приобретение автомобиля       автомобиль        однодетная  
2               покупка жилья            жилье         бездетная  
3  дополнительное образование      образование       многодетная  
4             сыграть свадьбу          свадьба         бездетная

<span style="color:blue">We form a pivot table based on the data of the children_category category debt value by which we want to see the pivot table.</span>

In [88]:
pd.pivot_table(data, index='children_category', values='debt')

debt
children_category          
бездетная          0.075129
малодетная         0.094404
многодетная        0.081579
однодетная         0.092154

### Сonclusion: Question 1

* Borrowers without children pay loans, only ≈ 7.5% do not pay loans.
* Borrowers from whom, many children pay loans and only 8.2% do not pay loans.
* Borrowers with one child pay loans and 9.2% do not pay loans.
* Borrowers with few children pay a loan and 9.4% are non-payers.

Thus, there is no direct correlation between the availability of children among borrowers and the repayment of a loan on time. Since the difference in the repayment of the loan and the number of debtors vary from 7.5% to 9.2%. Accordingly, 1.7% is the difference between the category in which there is a maximum number of credit defaults and the category in which there is a minimum number of non-paying borrowers. I think that these are not significant differences of 1.7%.
<br>
<br>
<br>

***

### Question 2
- **Is there a relationship between marital status and repayment of the loan on time?**

In order to find a relationship between the marital status of the borrower and repayment of the loan on time, it is necessary to segment the borrowers by the marital status of the borrowers.

That is, we need to categorize the data by the family_status and debt columns. <br>
So we categorize the data on the marital status of borrowers and the presence of debt to repay the loan.

<span style="color:blue">Distribute borrowers by marital status.</span>

In [89]:
def categorize_family_status(value):
    if value == 'женат / замужем':
        return 'в браке'
    else:   
        return 'не в браке'

<span style="color:blue">Applying the apply () data categorization method to values in the family_status column.</span>

In [90]:
data['family_status_category'] = data['family_status'].apply(categorize_family_status)

In [91]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose purpose_category children_category  \
0               покупка жилья            жилье        однодетная   
1     приобретение автомобиля       автомобиль        однодетная   
2               покупка жилья            жилье         бездетная   
3  дополнительное образование      образование       многодетная   
4             сыграть свадьбу          свадьба         бездетная   

  family_status_category  
0                в браке  
1                в браке  
2                в браке  
3                в браке  
4             не в браке

<span style="color:blue">Formed a pivot table based on the data of the category family_status_category debt value by which we want to see the pivot table.</span>

In [92]:
pd.pivot_table(data, index='family_status_category', values='debt')

debt
family_status_category          
в браке                 0.075354
не в браке              0.088462

### Сonclusion: Question 2

* Married borrowers repay loans, only ≈ 7.5% do not repay loans.
* Unmarried borrowers repay loans, only ≈ 8.8% do not repay loans.

Thus, it is interesting that married people more often pay loans and do not remain in debtors.
<br>
<br>
<br>

***

### Question 3
- **Is there a relationship between income and repayment of the loan on time?**

In order to find the relationship between the borrower's income level and the loan repayment on time, it is necessary to segment the borrowers by total income.

That is, we need to categorize the data by the column total_income and debt. <br>
So we categorize the data on the borrower's income and the presence of debt to repay the loan.

To answer the question about the relationship between the income level and repayment of loans on time, we need to categorize the data on the income of borrowers. To do this, use the distribution quantiles of 25%, 50% and 75%.

In [138]:
income_category = data['total_income'].describe()
income_category[4:7]

25%    107475.25
50%    143204.00
75%    198251.25
Name: total_income, dtype: float64

<span style="color:blue">We will distribute the borrowers by income level and repayment of the loan on time.<br>
We will write a function to determine the income category of borrowers.</span>

In [141]:
def categorize_total_income(income):
    if income <= income_category[4]: return 1
    elif income_category[4] < income <= income_category[5]: return 2
    elif income_category[5] < income <= income_category[6]: return 3
    else: return 4

<span style="color:blue">Applying the apply () data categorization method to values in the total_income column.</span>

In [12]:
data['total_income_category'] = data['total_income'].apply(categorize_total_income)
data.head()

NameError: name 'categorize_total_income' is not defined

In [143]:
data_pivot = data.pivot_table(index = ['total_income_category'], columns = 'debt', values = 'gender', aggfunc = 'count')

<span style="color:blue">Calculated the probability of debt for each income group.</span>

In [144]:
data_pivot['ratio'] = round(data_pivot[1] / (data_pivot[0] + data_pivot[1]), 3)
data_pivot.sort_values('ratio', ascending = False)

debt                      0    1  ratio
total_income_category                  
2                      4858  479  0.090
3                      4882  455  0.085
1                      4911  426  0.080
4                      4965  372  0.070

### Сonclusion: Question 3

The resulting debt probabilities:
* income in the range from 107.6 to 142.6 thousand rubles per month = 9.0%
* income in the range from 142.6 to 195.8 thousand rubles = 8.5%
* income less than or equal to 107.6 thousand rubles. = 8.0%
* more than 195.8 thousand rubles = 7.0%

* Borrowers with the highest income (more than 195.8 thousand rubles) are less likely to be in debtors, the percentage of loan debtors is ≈ 7.0%.
* Borrowers with a low level of income (income less than or equal to 107.6 thousand rubles), as well as borrowers with the highest income, try to pay off loans on time, so for borrowers from the group with the lowest income, the percentage of debtors on loans is ≈ 8 , 0%. This can be explained by the fact that people with the lowest incomes are better at managing their funds and are able to control their budget than those who earn a little more.
* Borrowers with an average income level from 107.6 to 142.6 thousand rubles a month remain in debt on loans with a percentage of ≈ 9.0%.
* Borrowers with an income level of 142.6 to 195.8 thousand rubles have, the percentage of debtors on loans ≈ 8.5%

Thus, loan repayment depends on the income level of borrowers. The higher the income of borrowers, the less% of loan arrears.

<br>
<br>

***

### Question 4
- **How do different loan objectives affect repayment on time?**

In order to find the relationship between the different goals of the borrowers and the repayment of the loan on time, it is necessary to segment the borrowers according to the purpose of the loan.

That is, we need to categorize the data by the purpose_category and debt columns. <br>
So we categorize the data on the goals of the borrowers and the presence of debt to repay the loan.

At the lemmatization stage, categories were created based on frequently occurring words:
* wedding
* the property
* housing
* car
* education

<span style="color:blue">We form a pivot table based on data from the category purpose_category value debt for which we want to see the pivot table.</span>

In [98]:
pd.pivot_table(data, index='purpose_category', values='debt')

debt
purpose_category          
автомобиль        0.093284
жилье             0.069213
недвижимость      0.074566
образование       0.092319
свадьба           0.078306

### Сonclusion: Question 4

* Borrowers with a loan obtained for the purchase of a car in 9.3% do not pay the loan;
* Borrowers with a loan received to pay for education in 9.2% do not pay the loan;
* Borrowers with a loan obtained for the purchase of housing in 6.9% do not pay the loan;
* Borrowers with a loan obtained for the acquisition of real estate in 7.5% do not pay the loan;
* Borrowers with a loan received to organize a wedding in 7.8% do not pay the loan;

Thus, the repayment of the loan depends on the goals of the borrowers. Depending on the purpose, you can predict a loan repayment.
<br>
<br>
<br>

***

## Step 4. General conclusion

Based on the data analysis, the following recommendations can be made to a bank for granting loans:
* The most often repayable loans issued for the purchase of housing, real estate and a wedding. There are the least debts on such loans.
* Loans issued for the purchase of a car and for the payment of education are not so often repaid.
* Credit repayment depends on the income level of borrowers. The higher the income of borrowers, the less% of loan arrears.
* Married people often pay loans and do not remain in debt.

<br>
<br>

***